### PR TITLE
adapted structure

### PR DESCRIPTION
**Pull request by kuegi**
rename to Bonds
extraacted BondToken as extra contract
added Ownable
added ReentrancyGuard
converted to errors instead of require

### CODE REVIEW

**Comment by kuegi** [comment]
this is based on feedback from the core team:

- Please use proper access control libraries from OpenZeppelin such as Ownable or AccessControl for all your smart contracts instead of using require(msg.sender == owner, ...). Side note, custom error/revert are more gas fee efficient rather than require() functions
- It's highly recommended not to extend any ERC721 contracts within the DUSDLock.sol contract for better access control and separation of concerns, to build upon this, a multi-sig wallet would be strongly recommended as the owner of the DUSDLock.sol contract as well for better role distribution. Best to deploy ERC721 contract separately and then call it from the DUSDLock.sol as an external call.
- Please consider using ReentrancyGaurd from OpenZeppelin as well, perhaps provide nonce per address interaction to prevent replay attacks.
- Natspec formatting would help other developers better understand your contracts and perhaps even contribute to it.

File: DUSDLock.sol [hunk 1]
```diff
@@ -86,92 +123,115 @@ contract DUSDLock is ERC721Enumerable {
...
-            uint256 batchId= tokenOfOwnerByIndex(msg.sender, idx);
+            uint256 batchId= bondToken.tokenOfOwnerByIndex(msg.sender, idx);
             uint256 claimed= availableRewards(batchId);
             if(claimed > 0) {
                 LockEntry storage entry= investments[batchId];
                 entry.claimedRewards += claimed;
                 total += claimed; 
                 emit RewardsClaimed(msg.sender, batchId, claimed, currentRewardsClaimable());
             }
+        }if(total == 0) {
```
Comment (Krysh90) [inline DUSDLock.sol:223]: missing new line

**Review by Krysh90 — APPROVED** [review]

**Review by Krysh90 — APPROVED** [review]

**Comment by 3DotsHub** [comment]
Quickly had a look over those changes. Looks good so far, however, need some more time for a deep dive.

**Comment by 3DotsHub** [comment]
Sadly I don't know so much about NFTs or ERC721, yet. And how to separate them to make external calls from the dusd-lock SC. Would need some time, for diving into those topics and play around a bit. Will keep you updated.

**Pull request merged**